tests/sub_port_interfaces/conftest.py

```diff
@@ -36,26 +43,34 @@ def define_sub_ports_configuration(request, duthost, ptfhost):
             }
         }
     """
+    sub_ports_config = {}
     vlan_ranges_dut = range(10, 30, 10)
     vlan_ranges_ptf = range(10, 30, 10)
 
     if 'invalid' in request.node.name:
         vlan_ranges_ptf = range(11, 31, 10)
 
-    interface_ranges = range(1, 2)
+    if 'max_numbers' in request.node.name:
+        vlan_ranges_dut = range(1, 257, 64)
+        vlan_ranges_ptf = range(1, 257, 64)
+
+        # Linux has the limitation of 15 characters on an interface name,
+        # but name of LAG port should have prefix 'PortChannel' and suffix
+        # '<0-9999>' on SONiC. So max length of LAG port suffix have be 3 characters
+        # For example: 'PortChannel1.96'
+        if request.param == 'port_in_lag':
+            vlan_ranges_dut = range(1, 97, 24)
+            vlan_ranges_ptf = range(1, 97, 24)
```

Comment:
Can you add comments elaborating why 256 sub-ports configuration is generated here?

Reply:
Linux has the limitation of 15 characters on an interface name and name of LAG port should have prefix 'PortChannel' and suffix '<0-9999>' on SONiC. So max length of sub-port suffix can be 4 characters  (such as, PortChannelX.XX, 15 characters). If we use just port, suffix can be 7 charecters (such as, EthernetXXX.XXX, 15 characters).

Reply:
For test case `test_max_numbers_of_sub_ports`, is it suppose to create totally 256 number of sub-ports, or just verify that sub-port with suffix like PortChannelX.XX can be created? 

I picked `define_sub_ports_configuration` and tested it from my side on t0. For test case with `max_numbers` in name, the dicts generated by this fixture are like below:
```
{
  "sub_ports": {
    "Ethernet4.129": {
      "ip": "172.16.0.9/30",
      "neighbor_port": "eth1.129",
      "neighbor_ip": "172.16.0.10/30"
    },
    "Ethernet8.129": {
      "ip": "172.16.4.9/30",
      "neighbor_port": "eth2.129",
      "neighbor_ip": "172.16.4.10/30"
    },
    "Ethernet8.65": {
      "ip": "172.16.4.5/30",
      "neighbor_port": "eth2.65",
      "neighbor_ip": "172.16.4.6/30"
    },
    "Ethernet4.1": {
      "ip": "172.16.0.1/30",
      "neighbor_port": "eth1.1",
      "neighbor_ip": "172.16.0.2/30"
    },
    "Ethernet4.65": {
      "ip": "172.16.0.5/30",
      "neighbor_port": "eth1.65",
      "neighbor_ip": "172.16.0.6/30"
    },
    "Ethernet4.193": {
      "ip": "172.16.0.13/30",
      "nei\r\nghbor_port": "eth1.193",
      "neighbor_ip": "172.16.0.14/30"
    },
    "Ethernet8.1": {
      "ip": "172.16.4.1/30",
      "neighbor_port": "eth2.1",
      "neighbor_ip": "172.16.4.2/30"
    },
    "Ethernet8.193": {
      "ip": "172.16.4.13/30",
      "neighbor_port": "eth2.193",
      "neighbor_ip": "172.16.4.14/30"
    }
  },
  "dut_ports": {
    "1": "Ethernet4",
    "2": "Ethernet8"
  },
  "ptf_ports": [
    "eth1",
    "eth2"
  ]
}
```
and
```
{
  "sub_ports": {
    "PortChannel1.1": {
      "ip": "172.16.0.1/30",
      "neighbor_port": "bond1.1",
      "neighbor_ip": "172.16.0.2/30"
    },
    "PortChannel2.73": {
      "ip": "172.16.4.\r\n13/30",
      "neighbor_port": "bond2.73",
      "neighbor_ip": "172.16.4.14/30"
    },
    "PortChannel2.49": {
      "ip": "172.16.4.9/30",
      "neighbor_port": "bond2.49",
      "neighbor_ip": "172.16.4.10/30"
    },
    "PortChannel1.25": {
      "ip": "172.\r\n16.0.5/30",
      "neighbor_port": "bond1.25",
      "neighbor_ip": "172.16.0.6/30"
    },
    "PortChannel2.25": {
      "ip": "172.16.4.5/30",
      "neighbor_port": "bond2.25",
      "neighbor_ip": "172.16.4.6/30"
    },
    "PortChannel1.49": {
      "ip": "172.16.0.9/30",
      "neighbor_port": "bond1.49",
      "neighbor_ip": "172.16.0.10/30"
    },
    "PortChannel1.73": {
      "ip": "172.16.0.13/30",
      "neighbor_port": "bond1.73",
      "neighbor_ip": "172.16.0.14/30"
    },
    "PortChannel2.1": {
      "ip": "172.16.4.1/30",
      "neighbor_port": "bond2.1",
      "neighbor_ip": "172.16.4.2/30"
    }
  },
  "dut_ports": {
    "1": "PortChannel1",
    "2": "PortChannel2"
  },
  "ptf_ports": {
    "bond1": "eth1",
    "bond2": "eth2"
  }
}
```

The total number of sub-ports is far less than 256. Is this intended?

Reply:
Test case `test_max_numbers_of_sub_ports` must create 256 sub-ports for port and 96 sub-ports for lag-port. Thanks for your help in investigating. I am working on this issue.

Reply:
This issue is fixed by changing steps of range in [block](https://github.com/Azure/sonic-mgmt/blob/48d084b3d6a5bdc3b301118ebe1bacb39ea60ddb/tests/sub_port_interfaces/conftest.py#L54-L55) (64 to 1), but the test needs too much time (~70 minutes) to run in this case. This behavior due to running of CLI commands for test setup and test teardown on the PTF and DUT. Does it make sense to change the steps of the range for that test case?

Reply:
Can you try the `shell_cmds` module? By using this module, you can put all the CLI commands to be executed in a list. Then use this `shell_cmds` module to run the commands one by one on remote host. It is more efficient than calling the shell module for each CLI command. Can you try it to see how much time is required then?

Reply:
I've tried the `shell_cmds` module. By using this module, the test case needs 40 minutes to run.

Reply:
Do we need to change the sub-ports range in this test case?

Reply:
I think 40 minutes is still too long. What about this idea:
* Add an option for specifying how many sub-ports to be covered. Value range of this option should be 1-256.
* Default value of the option can be much lesser to limit the test execution time.
Then we can run this test case by specifying value 256 for the option to have a full coverage when necessary. During daily regression, we can just cover the default range.

Reply:
I've added `--max_numbers_of_sub_ports` option for test_max_numbers_of_sub_ports case. Example of the command line how to run this tests:
`py.test --testbed=testbed-t0 --inventory=../ansible/lab --testbed_file=../ansible/testbed.csv --host-pattern=testbed-t0 -- module-path=../ansible/library sub_port_interfaces --max_numbers_of_sub_ports=100`